### PR TITLE
Tracking task for kangaroo lower body + 5dof

### PR DESCRIPTION
**Pull request by ManuelActisCassa**
Check workaround done in csv_to_npz script. It forced use of wandb, but commenting it all out is maybe not the ideal solution...

### CODE REVIEW

File: src/pal_mjlab/tasks/tracking/kangaroo/env_cfgs.py [hunk 1]
```diff
@@ -89,6 +93,26 @@ def pal_kangaroo_flat_tracking_env_cfg(
     "arm_right_tip_link",
   )
 
+  # ── Observations ──────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo/env_cfgs.py]:
```suggestion
  ## Observations
```

File: src/pal_mjlab/tasks/tracking/kangaroo/env_cfgs.py [hunk 2]
```diff
@@ -175,3 +199,22 @@ def pal_kangaroo_flat_tracking_env_cfg(
     motion_cmd.sampling_mode = "start"
 
   return cfg
+
+
+def pal_kangaroo_hands_flat_env_cfg(
+  has_state_estimation: bool = True,
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo/env_cfgs.py:205]:
```suggestion
  has_state_estimation: bool = False,
```

File: src/pal_mjlab/tasks/tracking/kangaroo/__init__.py [hunk 3]
```diff
@@ -21,3 +24,20 @@
...
 )
+
+
+register_mjlab_task(
+  task_id="Mjlab-Tracking-Flat-Pal-Kangaroo-Hands",
+  env_cfg=pal_kangaroo_hands_flat_env_cfg(has_state_estimation=True),
+  play_env_cfg=pal_kangaroo_hands_flat_env_cfg(has_state_estimation=False, play=True),
+  rl_cfg=pal_kangaroo_tracking_ppo_runner_cfg(),
+  runner_cls=MotionTrackingOnPolicyRunner,
+)
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo/__init__.py:35]:
```suggestion
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/__init__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+from .rl_cfg import pal_kangaroo_lower_body_tracking_ppo_runner_cfg
+
+register_mjlab_task(
+  task_id="Mjlab-Tracking-Flat-Pal-Kangaroo_Lower_Body",
+  env_cfg=pal_kangaroo_lower_body_flat_tracking_env_cfg(),
+  play_env_cfg=pal_kangaroo_lower_body_flat_tracking_env_cfg(play=True),
+  rl_cfg=pal_kangaroo_lower_body_tracking_ppo_runner_cfg(),
+  runner_cls=MotionTrackingOnPolicyRunner,
+)
+
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/__init__.py:14]:
```suggestion
```

Should we remove the state estimation one?

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    # Knee
+    "leg_left_knee_collision",
+    "leg_left_knee_bar_collision",
+    "leg_right_knee_collision",
+    "leg_right_knee_bar_collision",
+    # Pelvis
+    "pelvis_2_collision",
+  )
+
+  # ── Sensors ───────────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Sensors
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+      entity="robot",
+    ),
+    secondary=ContactMatch(mode="body", pattern="terrain"),
+    fields=("found",),
+    reduce="none",
+    num_slots=1,
+  )
+  cfg.scene.sensors = (self_collision_cfg, feet_ground_contact_cfg, body_ground_cfg)
+
+  # ── Actions ───────────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Actions
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+  joint_pos_action = cfg.actions["joint_pos"]
+  assert isinstance(joint_pos_action, JointPositionActionCfg)
+  joint_pos_action.scale = KANGAROO_LOWER_BODY_ACTION_SCALE
+  joint_pos_action.actuator_names = KANGAROO_LOWER_BODY_ACTUATOR_NAMES
+
+  # ── Commands ──────────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Commands
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    "pelvis_2_link",
+    "leg_left_3_link",
+    "leg_left_4_link",
+    "leg_left_5_link",
+    "leg_right_3_link",
+    "leg_right_4_link",
+    "leg_right_5_link",
+  )
+
+  # ── Rewards ───────────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Rewards
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+  cfg.rewards["foot_air"] = RewardTermCfg(
+    func=tracking_mdp.feet_air_time,
+    weight=2.0,
+    params={
+      "sensor_name": "feet_ground_contact",
+      "threshold": 0.1,
+    },
+  )
+
+  # ── Events (Domain Randomization) ─────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Events (Domain Randomization)
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+      "asset_cfg": SceneEntityCfg("robot", geom_names=body_geoms),
+      "operation": "abs",
+      "ranges": (0.3, 2.0),
+      "shared_random": False,
+    },
+  )
+
+  cfg.events["base_com"].params["asset_cfg"].body_names = ("pelvis_2_link",)
+
+  # ── Terminations ──────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Terminations
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+    func=mdp.illegal_contact,
+    params={"sensor_name": "body_ground_contact"},
+  )
+
+  cfg.terminations["ee_body_pos"].params["body_names"] = (
+    "leg_left_5_link",
+    "leg_right_5_link",
+  )
+
+  # ── Viewer ────────────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Viewer
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+  cfg.viewer.body_name = "base_link"
+
+  # ── Observations ──────────────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Observations
```

File: src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+      for k, v in cfg.observations["actor"].terms.items()
+      if k not in ["motion_anchor_pos_b", "motion_anchor_ori_b", "base_lin_vel"]
+    }
+    cfg.observations["actor"] = ObservationGroupCfg(
+      terms=new_actor_terms,
+      concatenate_terms=True,
+      enable_corruption=True,
+    )
+
+  # ── Play mode overrides ───────────────────────────────────────────────────
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/kangaroo_lower_body/env_cfgs.py]:
```suggestion
  ## Play mode overrides
```

File: src/pal_mjlab/tasks/tracking/mdp/rewards.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  cmd_foot_idxs = [command.cfg.body_names.index(name) for name in foot_body_names]
+  sensor_primary_names = list(
+    dict.fromkeys(slot.primary_name for slot in sensor._slots)
+  )
+  sensor_foot_idxs = [sensor_primary_names.index(name) for name in foot_body_names]
+
+  # Reference foot heights in world frame: [B, num_feet]
+  foot_heights = command.body_pos_w[:, cmd_foot_idxs, 2]
+  should_contact = foot_heights < height_threshold
+
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/mdp/rewards.py]: Can you add a test for this new reward?

File: src/pal_mjlab/tasks/tracking/mdp/rewards.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  sensor_name: str,
+  threshold: float = 0.02,
+) -> torch.Tensor:
+  sensor: ContactSensor = env.scene[sensor_name]
+  air_time = sensor.data.current_air_time  # [B, F]
+
+  # foot is considered airborne if it has been off contact long enough
+  in_air = air_time > threshold  # [B, F]
+
+  # both feet airborne → actual jump
```
Comment (saikishor) [inline src/pal_mjlab/tasks/tracking/mdp/rewards.py:40]: Same here

File: src/pal_mjlab/tasks/velocity/mdp/rewards.py [hunk 16]
```diff
@@ -494,3 +494,10 @@ def __call__(
 
     # return the squared sum of change in velocity along the projected gravity direction
     return cost
+
+
+def flat_reward(
+  env: ManagerBasedRlEnv,
+):
+  reward = torch.ones((env.scene.num_envs,), device=env.device)
+  return reward
```
Comment (saikishor) [inline src/pal_mjlab/tasks/velocity/mdp/rewards.py]: Same here ;)

**Review by saikishor — CHANGES_REQUESTED** [review]

**Review by saikishor — APPROVED** [review]
